### ISSUE TITLE
Speedup our geometry functions

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#2227]
## 🚀 Feature
@jatentaki reported that kornia functions are  clean, but slower that optimized ones. 
I have quickly checked and yes - for example, `batch_episym` from OANet (https://github.com/zjhthu/OANet/blob/master/core/loss.py#L5 ) is 2x faster on CPU than our implementation `K.geometry.symmetrical_epipolar_distance`. Our one is better on CUDA though :)

To reproduce:

```python3
import torch
import kornia as K

# From https://github.com/zjhthu/OANet/blob/master/core/loss.py#L5
def batch_episym(x1, x2, F):
    batch_size, num_pts = x1.shape[0], x1.shape[1]
    x1 = torch.cat([x1, x1.new_ones(batch_size, num_pts,1)], dim=-1).reshape(batch_size, num_pts,3,1)
    x2 = torch.cat([x2, x2.new_ones(batch_size, num_pts,1)], dim=-1).reshape(batch_size, num_pts,3,1)
    F = F.reshape(-1,1,3,3).repeat(1,num_pts,1,1)
    x2Fx1 = torch.matmul(x2.transpose(2,3), torch.matmul(F, x1)).reshape(batch_size,num_pts)
    Fx1 = torch.matmul(F,x1).reshape(batch_size,num_pts,3)
    Ftx2 = torch.matmul(F.transpose(2,3),x2).reshape(batch_size,num_pts,3)
    ys = x2Fx1**2 * (
            1.0 / (Fx1[:, :, 0]**2 + Fx1[:, :, 1]**2 + 1e-15) +
            1.0 / (Ftx2[:, :, 0]**2 + Ftx2[:, :, 1]**2 + 1e-15))
    return ys

B=16
N=1024

mps=torch.device('mps')
cpu=torch.device('cpu')

for device in [cpu, mps]:
    pt1 = torch.rand(B,N,2).float().to(device)
    pt2 = torch.rand(B,N,2).float().to(device)
    Fm = torch.rand(B,3,3).float().to(device)
    print ("OANet", device)
    %timeit epi_oanet = batch_episym(pt1, pt2, Fm)
    print ("kornia", device)
    %timeit epi_kornia = K.geometry.symmetrical_epipolar_distance(pt1, pt2, Fm)
```

```
OANet cpu
749 µs ± 3.84 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
kornia cpu
1.44 ms ± 10.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
OANet mps
1.24 ms ± 25.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
kornia mps
812 µs ± 4.32 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
OANet cuda
295 µs ± 32.8 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
kornia cuda
209 µs ± 4.16 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

That doesn't mean we should do smth immediately, but have a speed benchmark to start with, would be good.

**Comment by edgarriba** [comment]
A bit tangential but I reported this at some point https://discuss.pytorch.org/t/matmul-slow-for-small-matrices/168425/2
pytorch is not specially good with small matrices operations

**Comment by edgarriba** [comment]
If anyone is interested in this, feel free to re-open the ticket. Thanks!
